### PR TITLE
lsats and lextract

### PR DESCRIPTION
**Pull request by uvvpavel**

### CODE REVIEW

File: lib_xua/src/core/buffer/decouple/decouple.xc [hunk 1]
```diff
@@ -170,11 +170,7 @@ static inline void _send_sample_4(chanend c_mix_out, int ch)
         mult = multOutPtr[ch];
     }
     {h, l} = macs(mult, sample, 0, 0);
-    h <<= 3;
-#if (STREAM_FORMAT_OUTPUT_RESOLUTION_32BIT_USED == 1)
-    h |= (l >>29) & 0x7; // Note: This step is not required if we assume sample depth is 24bit (rather than 32bit)
-                            // Note: We need all 32bits for Native DSD
-#endif
+    asm("lextract %0,%1,%2,%3,32":"=r"(h):"r"(h),"r"(l),"r"(3));
```
Comment (xross) [inline lib_xua/src/core/buffer/decouple/decouple.xc:173]: Thanks for the update, can we leave the xc code in as comment for readability please, just so people who are not familiar with the lextract inst know what the intention is.

**Comment by xross** [comment]
My guess is the tests fail because the mixer routing uses self modifying code and the address offsets may now have changes, but I have not investigated at all.

File: lib_xua/src/core/buffer/decouple/decouple.xc (hunk 1 above)
Comment (xross) [reply lib_xua/src/core/buffer/decouple/decouple.xc:173]: (3) should be (32+3)?

Comment (uvvpavel) [reply lib_xua/src/core/buffer/decouple/decouple.xc:173]: Yes, it should be 29, I only understood that on a flight to my holiday 😆

Comment (uvvpavel) [reply lib_xua/src/core/buffer/decouple/decouple.xc:173]: @xross what part of the code do you want me to keep as a comment?
`h <<= 3; h |= (l >>29) & 0x7;` with some message about xs1 vs xs2?

**Comment by xross** [comment]
> My guess is the tests fail because the mixer routing uses self modifying code and the address offsets may now have changes, but I have not investigated at all.
> 
> The offset is hard-coded here:
> 
> https://github.com/xmos/lib_xua/blob/e1b41b4aed1f9ecd79bcefe36de9e0ed9f288225/lib_xua/src/core/mixer/fastmix.S#L150

@shuchitak reports this is not the issue, the offset doesn't change. Tests pass with 3 changed to 29 in the line reference above.

**Comment by uvvpavel** [comment]
Closing the issue as @shuchitak said she'll have to add a test for the 32-bit case and will incorporate these changes in her future work

**Pull request closed without merging**